### PR TITLE
Feature/update to fgbio

### PR DESCRIPTION
**Pull request by rhshah**
🚀  Adding a working version of the Fgbio bam generation workflow

### CODE REVIEW

**Comment by rhshah** [comment]
Need to add documentation before merging

**Comment by ionox0** [comment]
@rhshah this looks great, I can see the test passes so happy to merge whenever 

The one thing I noticed (which is really a separate issue) is that it would be great for us to have some way to keep these workflows from getting too large, so hopefully the CWL spec will follow through on this idea to allow `record` objects to be accessed without "parameter references" (such as `$(inputs.abra_params.sga)`)

This way we could use SchemaDefRequirement to group parameters together and not have to use so many separate inputs

https://cwl.discourse.group/t/accessing-record-fields-in-inputs-outputs-without-expressions/222

File: nucleo.cwl [hunk 1]
```diff
@@ -0,0 +1,810 @@
...
+    'sbg:y': 1281.9375
+  - id: gatk_base_recalibrator_known_sites
+    type:
+      type: array
+      items: File
+      inputBinding:
+        prefix: '--known-sites'
+    secondaryFiles:
+      - .idx
+    'sbg:x': 0
```
Comment (ionox0) [inline nucleo.cwl:37]: I can add the github action for removing these x and y tags if we need it

File: nucleo.cwl [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+      - id: abra2_contig_anchor
+        default: '10,1'
+        source: abra2_contig_anchor
+      - id: abra2_consensus_sequence
+        default: true
+        source: abra2_consensus_sequence
+      - id: gatk_merge_bam_alignment_output_file_name
+        source: BC_gatk_merge_bam_alignment_output_file_name
+      - id: fgbio_filter_consensus_read_reverse_per_base_tags_simplex_duplex
+        default: true
```
Comment (ionox0) [inline nucleo.cwl:732]: I'm not aware of this parameter and why it needs to be `true` but if it is necessary for the workflow to run I'm ok with it

File: nucleo.cwl [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+        source: BC_picard_addRG_output_file_name
+      - id: bwa_mem_output
+        source: BC_bwa_mem_output
+      - id: create_bam_index
+        default: true
+        source: create_bam_index
+      - id: bwa_mem_K
+        default: 1000000
+      - id: abra2_window_size
+        default: '800,700'
```
Comment (ionox0) [inline nucleo.cwl:695]: are there any parameters for `fastp` that need to have defaults here? or maybe we just use the tool defaults

**Review by ionox0 — APPROVED** [review]

**Comment by ionox0** [comment]
Sorry I should have thought of this before, but probably the better way for me to review this is to look at the travis logs. I picked out the commands and put them here:

https://gist.github.com/ionox0/6c1c0209648eb445386d51b08a383942 

The only potential issue that I see us that `--TMP_DIR /tmp` and `--env=TMPDIR=/tmp \` show up in a few places

Maybe this won't be an issue since it's the `/tmp` inside of the container, but we might want to make sure `/tmp` isn't being used anywhere. Logs about the mount points might also be an issue:
`--mount=type=bind,source=/tmp/4x8b9jiq,target=/jRYTeW`

But maybe this isn't an issue because when we set the TMPDIR environment variable set to /scratch these will all be changed. Just something to look out for.

**Comment by ionox0** [comment]
Also, after looking through the logs I found a few warnings that we may want to look into:

1. This is probably not an issue:
```
Skipping Docker software container '--memory' limit despite presence of ResourceRequirement with ramMin and/or ramMax setting. Consider running with --strict-memory-limit for increased portability assurance.
```

2. This may be not an issue, as it sounds like it corrects itself, but I'm just making sure that there isn't supposed to be a step before this that sorts the reads by queryname:
```
WARNING	2020-11-25 05:08:36	SamAlignmentMerger	Exception merging bam alignment - attempting to sort aligned reads and try again: Underlying iterator is not queryname sorted: K00121:236:HVYTTBBXX:3:2228:28980:16366 1/2 96b aligned to 14:19038550-19038645. > K00121:236:HVYTTBBXX:3:1101:3112:11319 1/2 95b aligned to 14:19041750-19041816.
```

3. It might be interesting to see the metrics that are produced when providing the reference sequence to this tool, I wonder if it could be a potential alternative for the noise calculation:
```
WARNING	2020-11-25 05:10:28	CollectAlignmentSummaryMetrics	Without a REFERENCE_SEQUENCE, metrics pertaining to mismatch rates will not be collected!
```
WARNING	2020-11-25 05:10:59	SamAlignmentMerger	Exception merging bam alignment - attempting to sort aligned reads and try again: Underlying iterator is not queryname sorted: test:9 2/2 95b aligned to 14:19487400-19487494. > test:6 1/2 96b aligned to 14:19487438-19487533.
```

**Comment by rhshah** [comment]
> Also, after looking through the logs I found a few warnings that we may want to look into:
> 
> 1. This is probably not an issue:
> 
> ```
> Skipping Docker software container '--memory' limit despite presence of ResourceRequirement with ramMin and/or ramMax setting. Consider running with --strict-memory-limit for increased portability assurance.
> ```
> 
> 1. This may be not an issue, as it sounds like it corrects itself, but I'm just making sure that there isn't supposed to be a step before this that sorts the reads by queryname:
> 
> ```
> WARNING	2020-11-25 05:08:36	SamAlignmentMerger	Exception merging bam alignment - attempting to sort aligned reads and try again: Underlying iterator is not queryname sorted: K00121:236:HVYTTBBXX:3:2228:28980:16366 1/2 96b aligned to 14:19038550-19038645. > K00121:236:HVYTTBBXX:3:1101:3112:11319 1/2 95b aligned to 14:19041750-19041816.
> ```
> 
> 1. It might be interesting to see the metrics that are produced when providing the reference sequence to this tool, I wonder if it could be a potential alternative for the noise calculation:
> 
> ```
> WARNING	2020-11-25 05:10:28	CollectAlignmentSummaryMetrics	Without a REFERENCE_SEQUENCE, metrics pertaining to mismatch rates will not be collected!
> ```

1. Correct
2. We have two files where one is `unmapped` sam file which we sort by `queryname` and second is `mapped` bam file which is `coordinate` sorted. Maybe if we change the output of the AddOrReplaceReadGroups to `queryname` instead of `coordinate` 
3. It might not be that useful as it will be an overall score rather than per base, but if we think it will be useful we can open it.

File: nucleo.cwl (hunk 2 above)
Comment (rhshah) [reply nucleo.cwl:732]: > I'm not aware of this parameter and why it needs to be `true` but if it is necessary for the workflow to run I'm ok with it

I think this is what @kanika-arora  had in her commands, I remember she gave an explanation but cannot recall, @kanika-arora  do you recall.

File: nucleo.cwl (hunk 1 above)
Comment (rhshah) [reply nucleo.cwl:37]: sure, please make sure it is done only when it is in development branch, not before that

File: nucleo.cwl (hunk 3 above)
Comment (rhshah) [reply nucleo.cwl:695]: @murphycj2  might know if we need to expose anything but I have set a couple of them as defaults

**Comment by rhshah** [comment]
> @rhshah this looks great, I can see the test passes so happy to merge whenever
> 
> The one thing I noticed (which is really a separate issue) is that it would be great for us to have some way to keep these workflows from getting too large, so hopefully the CWL spec will follow through on this idea to allow `record` objects to be accessed without "parameter references" (such as `$(inputs.abra_params.sga)`)
> 
> This way we could use SchemaDefRequirement to group parameters together and not have to use so many separate inputs
> 
> https://cwl.discourse.group/t/accessing-record-fields-in-inputs-outputs-without-expressions/222

You might need to explain me this a little probably we can have 1:1 to talk about this

**Comment by rhshah** [comment]
@kanika-arora and @murphycj2 if you guys can comment on things that @ionox0  brought up that will be great

File: nucleo.cwl (hunk 3 above)
Comment (murphycj2) [reply nucleo.cwl:703]: @rhshah @ionox0 Yes, based on the default parameters used for trimgalore, I set these parameters for fastp for a comparable trimming result:
- `--length_required 25`
- `--adapter_sequence GATCGGAAGAGC`
- `--adapter_sequence_r2 AGATCGGAAGAGC`
- `--disable_quality_filtering`

Comment (rhshah) [reply nucleo.cwl:717]: @murphycj2 thanks I updated the request to have this

**Pull request merged**